spock-spring/src/test/groovy/org/spockframework/spring/SharedFieldsInjection.groovy

```diff
@@ -0,0 +1,294 @@
+package org.spockframework.spring
+
+import org.spockframework.util.ReflectionUtil
+import org.springframework.test.context.ContextConfiguration
+import spock.lang.Specification
+import spock.util.EmbeddedSpecRunner
+
+import javax.inject.Inject
+
+import static org.springframework.test.annotation.DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD
+import static org.springframework.test.annotation.DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD
+
+class SharedFieldsInjection extends Specification {
+
+  EmbeddedSpecRunner runner = new EmbeddedSpecRunner()
+
+  def setup() {
+    runner.addClassImport(ContextConfiguration)
+    runner.addClassImport(Inject)
+    runner.addClassImport(EnableSharedInjection)
+    runner.addClassImport(IService1)
+  }
+
+  def "shared fields cannot be injected by default"() {
+    when:
+    runner.runWithImports """
+      @ContextConfiguration
+      class Foo extends Specification {
+        @$ann
+        @Shared
+        IService1 sharedService
+
+        def foo() {
+          expect: true
+        }
+      }
+    """
+
+    then:
+    def e = thrown(SpringExtensionException)
+    e.message == "@Shared field injection is not enabled by default therefore 'sharedService' field cannot be injected. " +
+      "Refer to javadoc of ${EnableSharedInjection.name} for information on how to opt-in for @Shared field injection."
+
+    where:
+    ann << [
+      "org.springframework.beans.factory.annotation.Autowired",
+      "javax.annotation.Resource",
+      "javax.inject.Inject"
+    ].findAll(ReflectionUtil.&isClassAvailable)
+  }
+
+  def "shared fields can be injected if opted-in for"() {
+    when:
+    runner.runWithImports '''
+      @ContextConfiguration(locations = "/org/spockframework/spring/InjectionExamples-context.xml")
+      @EnableSharedInjection
+      class Foo extends Specification {
+        @Inject
+        @Shared
+        IService1 sharedService
+
+        def setupSpec() {
+          assert sharedService != null
+        }
+
+        def test() {
+          expect:
+          sharedService != null
+        }
+      }
+    '''
+
+    then:
+    noExceptionThrown()
+  }
+
+  def "shared fields can be injected if opted-in for on a superclass"() {
```

Comment:
I'd like a more variations (A <- B <- C inherits)

1. Add `C` to this with simple inheritance
2. `@EnableSharedInjection` on `A`, but no fields, shared field and spring context in `B`, and simple inheritance in `C` 
3. `@EnableSharedInjection` on `A`, but no fields, shared field and spring context in `B`, and another shared field injection in `C` 
4.  `A` with shared field, `@EnableSharedInjection` on spring context in `B` -> should fail AFAICS

Reply:
I have to admit that I'm struggling to follow what kind of issues you are looking to test for with the suggested coverage. Can you please clarify the following for me so that I get a chance to provide exactly what you are after? In the above 4 cases:
- which, if any, specs should be abstract?
- which specs should have a `setupSpec()` and which a test that verifies that the injection worked?
- what failure are you expecting in `4.`? is that an expected failure and the test will just document it or is it due to a bug?

Thanks.